### PR TITLE
[MC][NFC] Use vector for GUIDProbeFunctionMap

### PR DESCRIPTION
**Pull request by aaupov**
Replace unordered_map with a vector. Pre-parse the section to statically
allocate storage. Use BumpPtrAllocator for FuncName strings, keep
StringRef in FuncDesc.

Reduces peak RSS of pseudo probe parsing from 9.08 GiB to 8.89 GiB as
part of perf2bolt with a large binary.

Test Plan:
```
bin/llvm-lit -sv test/tools/llvm-profgen
```

### CODE REVIEW

File: llvm/lib/MC/MCPseudoProbe.cpp [hunk 1]
```diff
@@ -390,32 +390,41 @@ bool MCPseudoProbeDecoder::buildGUID2FuncDescMap(const uint8_t *Start,
   Data = Start;
   End = Data + Size;
 
+  uint32_t FuncDescCount = 0;
   while (Data < End) {
-    auto ErrorOrGUID = readUnencodedNumber<uint64_t>();
-    if (!ErrorOrGUID)
+    if (!readUnencodedNumber<uint64_t>())
```
Comment (wlei-llvm) [inline llvm/lib/MC/MCPseudoProbe.cpp:395]: nit: comment the decoded field(and the one below)

File: llvm/include/llvm/MC/MCPseudoProbe.h [hunk 2]
```diff
@@ -100,8 +101,15 @@ class MCDecodedPseudoProbe;
...
 using MCPseudoProbeInlineStack = SmallVector<InlineSite, 8>;
 // GUID to PseudoProbeFuncDesc map
-using GUIDProbeFunctionMap =
-    std::unordered_map<uint64_t, MCPseudoProbeFuncDesc>;
+class GUIDProbeFunctionMap : public std::vector<MCPseudoProbeFuncDesc> {
+public:
+  auto find(uint64_t GUID) const {
+    return llvm::lower_bound(
+        *this, GUID, [](const MCPseudoProbeFuncDesc &Desc, uint64_t GUID) {
+          return Desc.FuncGUID < GUID;
```
Comment (wlei-llvm) [inline llvm/include/llvm/MC/MCPseudoProbe.h:109]: This is similar to the previous PR. we need to make sure return `end()` when GUID doesn't exist. (this case did happen before https://github.com/llvm/llvm-project/issues/98127)

File: llvm/lib/MC/MCPseudoProbe.cpp [hunk 3]
```diff
@@ -390,32 +390,41 @@ bool MCPseudoProbeDecoder::buildGUID2FuncDescMap(const uint8_t *Start,
...
+    StringRef Name = cantFail(errorOrToExpected(readString(NameSize)));
 
     // Initialize PseudoProbeFuncDesc and populate it into GUID2FuncDescMap
-    GUID2FuncDescMap.emplace(GUID, MCPseudoProbeFuncDesc(GUID, Hash, Name));
+    GUID2FuncDescMap.emplace_back(GUID, Hash, Name.copy(FuncNameAllocator));
   }
   assert(Data == End && "Have unprocessed data in pseudo_probe_desc section");
+  llvm::sort(GUID2FuncDescMap, [](const auto &LHS, const auto &RHS) {
+    return LHS.FuncGUID < RHS.FuncGUID;
+  });
```
Comment (wlei-llvm) [inline llvm/lib/MC/MCPseudoProbe.cpp:427]: Similarly, add the assertion for the reserved size.

**Review by wlei-llvm — APPROVED** [review]
LGTM, thanks.

**Comment by cumcum8197** [comment]
a79cf0228e4ade1f69e8196fdcf9c0184b4ba1b7

**Pull request merged**